crates/mun_hir/src/name_resolution/path_resolution.rs

```diff
@@ -0,0 +1,149 @@
+use crate::ids::{ItemDefinitionId, ModuleId, PackageId};
+use crate::item_scope::BUILTIN_SCOPE;
+use crate::module_tree::LocalModuleId;
+use crate::package_defs::PackageDefs;
+use crate::{DefDatabase, Name, Path, PathKind, PerNs, Visibility};
+use itertools::__std_iter::successors;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(super) enum ReachedFixedPoint {
+    Yes,
+    No,
+}
+
+/// Contains the result of resolving a path. It contains how far the path was able to be resolved
+/// as well as the resolved values or types so far.
+#[derive(Debug, Clone)]
+pub(super) struct ResolvePathResult {
+    pub(super) resolved_def: PerNs<(ItemDefinitionId, Visibility)>,
+    pub(super) segment_index: Option<usize>,
+    pub(super) reached_fixedpoint: ReachedFixedPoint,
+    pub(super) package: Option<PackageId>,
+}
+
+impl ResolvePathResult {
+    /// Constructs an empty `ResolvePathResult`
+    fn empty(reached_fixedpoint: ReachedFixedPoint) -> ResolvePathResult {
+        ResolvePathResult::with(PerNs::none(), reached_fixedpoint, None, None)
+    }
+
+    /// Constructs a new instance of `ResolvePathResult`
+    fn with(
+        resolved_def: PerNs<(ItemDefinitionId, Visibility)>,
+        reached_fixedpoint: ReachedFixedPoint,
+        segment_index: Option<usize>,
+        package: Option<PackageId>,
+    ) -> ResolvePathResult {
+        ResolvePathResult {
+            resolved_def,
+            reached_fixedpoint,
+            segment_index,
+            package,
+        }
+    }
+}
+
+impl PackageDefs {
+    /// Resolves the specified `path` from within the specified `module`. Also optionally returns
+    /// which part of the path was resolved, if this is not `None` it means the path didn't resolve
+    /// completely yet.
+    pub(crate) fn resolve_path_in_module(
+        &self,
+        db: &dyn DefDatabase,
+        module: LocalModuleId,
+        path: &Path,
+    ) -> (PerNs<(ItemDefinitionId, Visibility)>, Option<usize>) {
+        let res = self.resolve_path_with_fixedpoint(db, module, path);
+        (res.resolved_def, res.segment_index)
+    }
+
+    /// Resolves the specified `name` from within the specified `module`
+    fn resolve_name_in_module(
+        &self,
+        _db: &dyn DefDatabase,
+        module: LocalModuleId,
+        name: &Name,
+    ) -> PerNs<(ItemDefinitionId, Visibility)> {
+        self[module]
+            .get(name)
+            .or(BUILTIN_SCOPE.get(name).copied().unwrap_or_else(PerNs::none))
+    }
+
+    /// Resolves the specified `path` from within the specified `module`. Also returns whether or
+    /// not an additions to the PackageDef would change the result. This is useful when resolving
+    /// all imports.
+    pub(super) fn resolve_path_with_fixedpoint(
```

Comment:
Based on the description, I'm not sure why this function has `fixedpoint` in its name.

Reply:
I modified the comment so that this is a bit clearer.